### PR TITLE
Llama2 Chatbot on Mac

### PR DESCRIPTION
**Pull request by agunapal**
## Description

This is an example showing how to deploy a llama2 chat app using TorchServe on your laptop!

We use [streamlit](https://github.com/streamlit/streamlit) to create the app

We are using [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) in this example

This example doesn't include streaming response. 
Though i was able to get the server to send the response, the client was receiving junk .

architecture

Server
Client

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: examples/LLM/llama2/chat_app/Readme.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/Readme.md:1]: mention this on the main README as well as a way for people to deploy an end to end chatbot

File: examples/LLM/llama2/chat_app/model-config.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# TorchServe frontend parameters
+responseTimeout: 1200
+
+handler:
+    model_name: "llama-cpp"
+    model_path: "/Users/agunapal/Documents/experiments/llama/ggml-model-q4_0.gguf"
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/model-config.yaml:6]: update this path or error in handler if its not set

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import requests
+import streamlit as st
+
+MODEL_NAME = "llamacpp"
+# App title
+st.set_page_config(page_title="🦙💬 Llama 2 TorchServe Serve")
+
+
+def start_server():
+    os.system("torchserve --start --model-store model_store --ncs")
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:13]: for snappier starts you can disable the compression in archiver

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    st.session_state.started = True
+    st.session_state.stopped = False
+    st.session_state.registered = False
+
+
+def stop_server():
+    os.system("torchserve --stop")
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:20]: add some simple log statements for each of hte main functions

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        return
+    st.session_state.registered = True
+    st.session_state.started = False
+    st.session_state.stopped = False
+    server_state_container.caption(res.text)
+
+
+def register_model():
+    if not st.session_state.started:
+        server_state_container.caption("TorchServe has stopped. Start it")
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:40]: confusing sentence

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+
+if "started" not in st.session_state:
+    st.session_state.started = False
+if "stopped" not in st.session_state:
+    st.session_state.stopped = False
+if "registered" not in st.session_state:
+    st.session_state.registered = False
+
+# Replicate Credentials
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:138]: wdym by credentials

File: examples/LLM/llama2/chat_app/Readme.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+#### Dynamic Batching
+
+You can make use of dynamic batching in TorchServe by configuring the `batch_size` and `max_batch_delay` parameters in TorchServe. You can do this on the Server app.
+
+![Batch Size](./screenshots/batch_size.png)
+
+#### Backend Workers
+
+You can increase the number of backend workers in TorchServe by configuring  `min_workers` parameter in TorchServe. You can do this on the Server app.
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/Readme.md:121]: clarify that workers refers to more processes

File: examples/LLM/llama2/chat_app/Readme.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+### Experiments
+
+You can launch a second client app from another terminal.
+
+You can send requests simultaneously to see how quickly TorchServe responds
+
+#### Dynamic Batching
+
+You can make use of dynamic batching in TorchServe by configuring the `batch_size` and `max_batch_delay` parameters in TorchServe. You can do this on the Server app.
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/Readme.md:115]: briefly explain waht dynamic batching is or provide a link to some reference

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    st.title("🦙💬 Llama 2 TorchServe Server ")
+
+    st.button("Start Server", on_click=start_server)
+    st.button("Stop Server", on_click=stop_server)
+    st.button("Register Llama2", on_click=register_model)
+    workers = st.sidebar.slider(
+        "Num Workers",
+        key="Num Workers",
+        min_value=1,
+        max_value=4,
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:149]: why max it out like this? I think you can make it up to the number of cores. If it crashes mgiht be cool to show how going to fewer workers also gives you error recovery

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        value=1,
+        step=1,
+        on_change=scale_workers,
+        args=("Num Workers",),
+    )
+    batch_size = st.sidebar.select_slider(
+        "Batch Size",
+        key="Batch Size",
+        options=[2**j for j in range(0, 8)],
+        on_change=set_batch_size,
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:159]: so how does batch size work here exactly? Like I'm typing things in chat does the server then respond multiple times at once?

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        st.success("Stopped TorchServe", icon="✅")
+
+    if st.session_state.registered:
+        st.success("Registered model", icon="✅")
+
+st.title("TorchServe Status")
+server_state_container = st.container()
+server_state_container.subheader("Server status:")
+
+if st.session_state.started:
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:186]: can the server fail after this point?

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    )
+    _register_model(url)
+
+
+def get_status():
+    if st.session_state.registered:
+        url = "http://localhost:8081/models/" + MODEL_NAME
+        res = requests.get(url)
+        if res.status_code != 200:
+            model_state_container.error("Error getting model status", icon="🚫")
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:58]: so hopefully actual error logs are still somewhere? Don't wanna swallow the actual error message for someone trying to debug this

File: examples/LLM/llama2/chat_app/torchserve_server_app.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+def set_batch_size(batch_size):
+    if st.session_state.registered:
+        url = "http://localhost:8081/models/" + MODEL_NAME + "/1.0"
+        res = requests.delete(url)
+        server_state_container.caption(res.text)
+        st.session_state.registered = False
+
+        batch_size = st.session_state[batch_size]
+        url = (
```
Comment (msaroufim) [inline examples/LLM/llama2/chat_app/torchserve_server_app.py:89]: f strings are nicer

**Review by msaroufim — APPROVED** [review]
Cool! a bunch of minor feedback but otherwise this is looking good

File: examples/LLM/llama2/chat_app/model-config.yaml (hunk 2 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/model-config.yaml:6]: I have taken care of it. I am setting an env variable and using that. But let me add a comment saying that users should updated this path in real usecase https://github.com/pytorch/serve/pull/2618/files#diff-f8d92d4ecdf7cd36b36149d88f699c651f5f1da2dc874ca13501bf524b1f6968R29

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 3 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:13]: Hmmm..I am not sure I follow.  When archiver is just passing the path to the weights using the yaml file and not the actual weights, this wouldn't come into picture?

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 6 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:138]: oops..copy paste :D

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 9 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:149]: Good point . Let me try and update the code/readme with description of what users can expect

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 10 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:159]: For each batch size, we will unregister the model and register with the new batch size. The app doesn't respond till the previous batch_size setup is completely executed. So, its pretty neat and you can't break it by moving the slider around

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 12 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:58]: Yes, the terminal where you start the server shows the actual logs. But let me add a comment in the readme

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 4 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:20]: The app has statements like these `server_state_container.success("Started TorchServe", icon="✅")` These show up on the app

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 5 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:40]: updated

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 6 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:138]: done

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 5 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py]: done

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 13 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:79]: totally forgot. Thanks!

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 4 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:20]: This exists in the app

File: examples/LLM/llama2/chat_app/torchserve_server_app.py (hunk 9 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/torchserve_server_app.py:120]: Seems to be running fine. My laptop is slow but nothing has crashed
Screenshot 2023-09-29 at 2 00 57 PM

File: examples/LLM/llama2/chat_app/Readme.md (hunk 1 above)
Comment (agunapal) [reply examples/LLM/llama2/chat_app/Readme.md:1]: added

**Pull request merged**